src/lib/coreaudio/CoreAudioDevice.cxx

```diff
@@ -0,0 +1,676 @@
+	score += (int_portion == 1.0) ? 500 : 0;
+	// prefer higher multiples if source rate higher than dest rate
+	if(source_rate >= destination_rate)
+		score += (int_portion > 1 && int_portion < 100) ? (100 - int_portion) / 100 * 100 : 0;
+	else
+		score += (int_portion > 1 && int_portion < 100) ? (100 + int_portion) / 100 * 100 : 0;
+	
+	return score;
+}
+
+float
+CoreAudioDevice::ScoreFormat(const AudioStreamBasicDescription &format_desc, const AudioFormat &format) const
+{
+	float score = 0;
+	// Score only linear PCM formats (everything else MPD cannot use)
+	if (format_desc.mFormatID == kAudioFormatLinearPCM)
+	{
+		score += ScoreSampleRate(format_desc.mSampleRate, format.sample_rate);
+		
+		// Just choose the stream / format with the highest number of output channels
+		score += format_desc.mChannelsPerFrame * 5;
+		
+		if (format.format == SampleFormat::FLOAT)
+		{
+			// for float, prefer the highest bitdepth we have
+			if (format_desc.mBitsPerChannel >= 16)
+				score += (format_desc.mBitsPerChannel / 8);
+		}
+		else
+		{
+			if (format_desc.mBitsPerChannel == ((format.format == SampleFormat::S24_P32) ? 24 : format.GetSampleSize() * 8))
+				score += 5;
+			else if (format_desc.mBitsPerChannel > format.GetSampleSize() * 8)
+				score += 1;
+		}
+	}
+	return score;
+}
+
+AudioStreamBasicDescription CoreAudioDevice::GetPhysFormat()
+{
+	/** Always report back the (selected) physical
+	 *	stream output format that will be set when
+	 *	setting up the IOProc.
+	 */
+	return output_format;
+}
+
+AudioStreamBasicDescription CoreAudioDevice::GetIOFormat()
+{
+	/** Return the physical format of the device in case integer
+	 *	mode is active (a non-mixable format has been selected).
+	 *	Otherwise, the virtual format is 32bit native endian float
+	 *	samples with the same sample rate and channels as the
+	 *	physical format.
+	 */
+	AudioStreamBasicDescription io_format;
+	if(output_format.mFormatFlags & kAudioFormatFlagIsNonMixable)
+		return output_format;
+	memset(&io_format, 0, sizeof(io_format));
+	io_format.mFormatID = kAudioFormatLinearPCM;
+	io_format.mChannelsPerFrame = output_format.mChannelsPerFrame;
+	io_format.mSampleRate = output_format.mSampleRate;
+	io_format.mFramesPerPacket = 1;
+	io_format.mFormatFlags = kAudioFormatFlagIsFloat;
+	if(IsBigEndian())
+		io_format.mFormatFlags |= kAudioFormatFlagIsBigEndian;
+	io_format.mBitsPerChannel = 32;
+	io_format.mBytesPerFrame = io_format.mBytesPerPacket = sizeof(Float32) * io_format.mChannelsPerFrame;
+	return io_format;
+}
+
+bool CoreAudioDevice::SetFormat(const AudioFormat &audio_format, bool prefer_unmixable)
+{
+	FormatDebug(macos_output_domain, "Finding CoreAudio stream for format %s.", ToString(audio_format).c_str());
+	
+	bool format_found  = false;
+	memset(&output_format, 0, sizeof(output_format));
+	float output_score  = 0;
+	
+	// loop over all streams to find one with a suitable format
+	for(UInt32 stream_idx = 0; stream_idx < stream_infos.size(); stream_idx++)
```

Comment:
Use range-based `for`

Reply:
Done

Reply:
Need to stick with traditional for loop as I need an index (count of the element in the vector)